### ISSUE TITLE
Suggestion: Attempt to whitelist self with HID Guardian during boot

### ISSUE THREAD

**Issue by Tahvohck** [WhiteMagic/JoystickGremlin#242]
At the moment vJoy + Joystick Gremlin + HID Guardian is pretty much the best combination I could ask for, but one thing really bothers me currently, and it's that I have to separately whitelist Gremlin after launching it. If Gremlin could attempt to whitelist itself during boot, even without admin privileges, that would make it so I only ever have to launch one program (Gremlin) before launching a game. I know that this can be done via the v1 API of HG:

`http://localhost:26762/api/v1/hidguardian/whitelist/add/<PID>`
`http://localhost:26762/api/v1/hidguardian/whitelist/remove/<PID>`

I'd submit a pull request, but I'm not sure where in the program flow would be best. The code would be pretty simple:

```python
import requests, os
pidadd = "http://localhost:26762/api/v1/hidguardian/whitelist/add/{}"
try: requests.get( pidadd.format(os.getpid()) )
except: pass
```

If this was done before device polling, it would also mean that the devices would be immediately available.

**Comment by WhiteMagic** [comment]
There hasn't been much additional work with regards to interfacing with HidGuardian, mainly because I was waiting for a new more refined version to be released. However, it seems that for the foreseeable future there won't be anything happening on that front.

The issue with using HidCerberus is that this works under the assumption that this has been installed alongside HidGuardian, which may not be the case. An additional problem is that Gremlin actually does a bit more then just whitelisting it's own PID. Gremlin annotates the PIDs that it added such that it can remove those upon launch to get rid of old PIDs that no longer are active. This would not be possible (at least directly) by using the HidCerberus route.

With regards to devices being available, this should already be the case, HidGuardian related actions happen before device listing etc. happens. If devices don't show up upon launch when HidGuardian is used with Gremlin launching as admin then there's something else wrong.

So overall I'm not sure if it's worth replacing the existing system for one that requires another system service to run. It would be nice not having to run Gremlin as root, however, I also don't know how many people have HidGuardian installed without HidCerberus.

**Comment by Tahvohck** [comment]
That's part of why this can be done in a try block. The program can confirm that the whitelist has one of its old PIDs via the v1 API call of:
`http://localhost:26762/api/v1/hidguardian/whitelist/get`
If launched in admin mode it can even bypass the Cerberus V1 api and immediately default to the direct registry interface. One of my gripes about this is that I believe in admin/user separation after a long while of using linux, and so I rankle any time a userspace program wants admin rights.

Keep in mind that I'm not suggesting replacing the existing system, only adding a fallback system alongside it for when the program is run without admin rights. If you can point me in the right direction I'll even fix up some code to do so, more than the psuedocode I put together for the initial post.

**Comment by WhiteMagic** [comment]
The actual code that handles interfacing with HidGuardian lives in https://github.com/WhiteMagic/JoystickGremlin/blob/develop/gremlin/hid_guardian.py. The whitelisting of Gremlin itself is only part of what is needed to support HidGuardian via HidCerberus' REST API though. I personally don't like having to run Gremlin as root, which is why I nearly didn't add support for HidGuardian as messing directly with the registry is really bad.

Effectively the whole set of methods of the HidGuardian class would need to be mapped onto the REST API and two separate classes with the same interface would need to exist, one going via the registry and the other one via HidCerberus. That way the current HidGuardian class would stay the same and when created would simply check for the most appropriate class implementing the functions and use that.

I'm also not sure how necessary the requests package is, while it's not too big, I personally dislike adding tons of dependencies when I only need a single trivial function of them.

**Comment by Tahvohck** [comment]
I'll look into that code and author a pull request. Regarding requests itself, it's not actually necessary I just don't know how to handle raw http requests in pure python and wanted to get a quick example code. Requests is good, but obviously no reason to use it if we don't need it.

Regarding messing with the registry, I've talked to Nefarius and their stance is that registry access was never supposed to be the way to interact with HG. The v1 API was RESTful, but I know nothing about v3, and sadly Nefarius isn't in a position to continue dev on HIDG at the moment. It may be better to just say that users that expect HIDG functionality need to have a user-mode interface installed as well (currently HIDC). Rolling back current support is probably the extreme option though.

Regarding full duplication, that may not be needed since HIDC provides an interface for hiding and unhiding devices already (`http://localhost:26762`), so the self-whitelisting is all that ACTUALLY needs to be done. On the other hand, no sense in doing things by half-measures.

**Comment by WhiteMagic** [comment]
Yeah the webinterface allows doing all these things quite a few people explicitly want everything to be in one tool, which is really the only reason Gremlin even got support for configuring HidGuardian. As for the registry never having been the intended way to configure it, that's to be expected and I didn't assume it was the intended way but it allowed me to do the auto cleanup bit. Yeah at this stage using v1 is the only real option for HidGuardian from what I can tell all other version are not really available and not intended to be used.

The bare metal http request stuff in Python is urllib https://docs.python.org/3/library/urllib.request.html#module-urllib.request I have not used this so I can't say how much of a pain this is. If it turns out that it's a nightmare and you have to write a ton of code just to get anything out of that then that's a good enough reason to use requests.

**Comment by Tahvohck** [comment]
So far so good on this. ~urllib2~ urllib hasn't been too hard to work with - I'd not use it large scale ever but for simple get requests so far it's been easy.

**Comment by muchimi** [comment]
Just to add on what I think is happening to me:  With the new version of windows 10, when running HID guardian, the first time I run Gremlin (in admin mode), it only sees the VJOY devices.

HID guardian shows gremlin as the proper "authorized" process using WhiteKnight's util.

I need to disconnect one physical controller for Gremlin to "wake up" and see all devices.

My current work-around is I got an inline physical USB switch (that works on data and power - power alone won't work) - so I throw the switch on/off/on to get Gremlin to wake up the first time I run it.

It didn't do that prior to Windows 10 1903.

Not sure what changed, but clearly, something's not normal.

**Comment by WhiteMagic** [comment]
Give this version a go https://my.pcloud.com/publink/show?code=XZzBCG7ZYQqx7BCGr5QH90srhAocxLFxpMAk this should fix the issue without breaking the new vJoy version.

**Comment by muchimi** [comment]
Works like a champ!  Thanks.  What was the issue?

**Comment by WhiteMagic** [comment]
Order of execution. Gremlin did initialize the devices before whitelisting itself with HidGuardian which meant it didn't get to see any device. An older unreleased version fixed this ordering issue but didn't include the fix that allowed for newer vJoy version being used so that's why I had to create one version that combined all the fixes together.

**Comment by muchimi** [comment]
Got it, thanks for the explanation.  I think the new version will solve a few issues for folks on the new version of windows.

**Comment by dixonte** [comment]
Have you considered creating a windows service for this task? You could have it monitor process creation via WMI, when joystick gremlin is detected it could add the PID to the whitelist, then signal joystick gremlin to refresh the device list via a named pipe. This would allow joystick gremlin to interact with HidGuardian without having admin rights itself.

I have some experience with windows services so I could contribute some code there, though it would be written in C#. But I'm not familiar enough with python to know how easy or not it would be to receive a signal via named pipe or other IPC method.

**Comment by WhiteMagic** [comment]
Not really no, the stuff muchimi is working on solves this effectively without having to add yet another thing. Installing HidGuardian via the tool by evilC will also add the HidCerberus web service which can be used via the REST API which doesn't require admin rights.

**Comment by dixonte** [comment]
I've tried evilC's tool. It is really useful, but I'm not a fan of HidCerberus. It runs as Local System, which effectively gives it admin rights, and opens a web endpoint while doing so. Granted it seems to be bound to local host, but still that's a concern. On top of which the version I saw didn't completely work (couldn't submit HID strings).

I'm going to keep poking around with my own solution. Reading more comments here, I think I can do it without needing any changes in Joystick Gremlin. We'll see, anyway.

Edit: I have a proof of concept, if you're interested: https://github.com/dixonte/JoystickGremlinWhitelister

**Comment by muchimi** [comment]
For reference, I use this tool to manage HID Guardian and HID Cerberus and set this up with Gremlin. 
 
https://www.autohotkey.com/boards/viewtopic.php?t=34890

The security question is an excellent one.  As far as I can tell, the webserver runs only on localhost using the built-in service account, and unless your ports are enabled on your local firewall - no-one should be able to get into your box.  You should be able to lock that down pretty fast.   An improvement would be to run the service using a service account you setup with more limited rights than the built-in account - however windows will require some pretty significant rights to get to hardware information.  

The current solution (load order) works perfectly for me right now in Gremlin.

**Comment by dixonte** [comment]
HidCerberus is already not available from outside your local machine since its port is bound to localhost, I'm pretty sure. Even without a firewall, I don't believe people on the internet or even the local LAN would be able to access it. I'm more concerned about the fact that any unprivileged program running locally can feed input into this thing.

Maybe I'm worrying over nothing. It was a fun exercise to make my own solution, too. I might polish it up a bit still and offer it as an alternative. The way I have it set up now I can just have Joystick Gremlin start with Windows, without admin rights, and everything just works without a single UAC prompt.

**Comment by dixonte** [comment]
Here's a more polished solution to the proof-of-concept I offered earlier, including a basic installer so you don't have to compile it: https://github.com/dixonte/HidVanguard

**Comment by WhiteMagic** [comment]
HidGuardian being deprecated in favor of HidHide and R14 dropping the built-in management for HidHide makes this redundant.
